### PR TITLE
Add tarred dataset support for NMT

### PR DESCRIPTION
**Pull request by MaximumEntropy**
This PR adds support for Tarred Datasets when training NMT models on very large datasets. It avoids loading the entire dataset into CPU memory for every DDP process. It uses webdataset (https://github.com/tmbdev/webdataset) similar to ASR and language modeling.

To use this functionality

1. Run `python examples/nlp/machine_translation/create_tarred_parallel_dataset.py --clean --bpe_dropout 0.1 --shared_tokenizer --src_fname src_file.txt --tgt_fname tgt_file.txt --out_dir /path/to/out_dir --vocab_size 32000 --tokens_in_batch 16000 --lines_per_dataset_fragment 5000000 --num_batches_per_tarfile 1000` to create a folder that contains many tar files. Each tar file consists of a certain number of batches (specified by `num_batches_per_tarfile`), where each batch is a pickle file. The output folder will also contain `metadata.json` that contains the total number of training batches and YTTM tokenizer model files.
2. Point to the tarred dataset folder when training like so

```
python examples/nlp/machine_translation/enc_dec_nmt.py -c aayn_base \
     trainer.gpus=8 \
     +trainer.max_steps=300000
     ........... 
     model.train_ds.src_file_name=/path/to/out_dir/batches.tokens.16000._OP_1..100_CL_.tar \
     model.train_ds.tgt_file_name=/path/to/out_dir/batches.tokens.16000._OP_1..100_CL_.tar \
     +model.train_ds.load_from_tarred_dataset=true \
     +model.train_ds.metadata_path=/path/to/out_dir/metadata.json \
     +model.train_ds.tar_shuffle_n=100 \
     .............
```

### CODE REVIEW

File: nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py [hunk 1]
```diff
@@ -14,22 +14,29 @@
...
-__all__ = ['TranslationDataset']
+__all__ = ['TranslationDataset', 'TarredTranslationDataset']
 
 
 @dataclass
 class TranslationDataConfig:
-    src_file_name: str = MISSING
-    tgt_file_name: str = MISSING
+    src_file_name: str = None  # MISSING
+    tgt_file_name: str = None  # MISSING
```
Comment (ericharper) [inline nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py:39]: Why `None` here?

File: nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py [hunk 2]
```diff
@@ -260,3 +269,148 @@ def clean_src_and_target(
             src_ids_.append(src_ids[i])
             tgt_ids_.append(tgt_ids[i])
         return src_ids_, tgt_ids_
+
+
+class TarredTranslationDataset(IterableDataset):
+    """
+    A similar Dataset to the TranslationDataset, but which loads tarred tokenized pickle files.
+    Accepts a single JSON metadata file containing the total number of batches
+    as well as the path(s) to the tarball(s) containing the wav files.
```
Comment (ericharper) [inline nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py:278]: wav files?

**Review by ericharper — APPROVED** [review]
Other than a couple minor changes, LGTM.

File: nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py [hunk 3]
```diff
@@ -260,3 +270,148 @@ def clean_src_and_target(
             src_ids_.append(src_ids[i])
             tgt_ids_.append(tgt_ids[i])
         return src_ids_, tgt_ids_
+
+
+class TarredTranslationDataset(IterableDataset):
+    """
+    A similar Dataset to the TranslationDataset, but which loads tarred tokenized pickle files.
```
Comment (okuchaiev) [inline nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py:277]: Nice docstring!

**Review by okuchaiev — APPROVED** [review]
lgtm

**Review by titu1994 — APPROVED** [review]
LGTM

**Pull request merged**